pkg/controllers/node/node_controller.go

```diff
@@ -38,46 +41,160 @@ const (
+		handlers.AddFunc = func(obj interface{}) {
+			nc.syncNodeLabels(obj.(*v1.Node))
+		}
+		handlers.UpdateFunc = func(_, obj interface{}) {
+			nc.syncNodeLabels(obj.(*v1.Node))
+		}
+	}
+
+	// Informer handles managing the watch and signals us when nodes are deleted.
+	// also syncs up labels between k8s/calico node objects
+	nc.indexer, nc.informer = cache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, handlers, cache.Indexers{})
+
+	return nc
+}
+
+// syncNodeLabels syncs the labels found in v1.Node to the Calico node object.
+// It uses an annotation on the Calico node object to keep track of which labels have
+// beend synced from Kubernetes, so that it doesn't overwrite user provided labels (e.g.,
+// via calicoctl or another Calico controller).
+func (nc *NodeController) syncNodeLabels(node *v1.Node) {
+	// On failure, we retry a certain number of times.
+	for n := 1; n < maxAttempts; n++ {
+		// Get the Calico node representation.
+		name, ok := nc.nodemapper[node.Name]
+		if !ok {
+			// We havent learned this Calico node yet.
+			log.Debugf("Skipping update for node with no Calico equivalent")
+			return
+		}
+		calNode, err := nc.calicoClient.Nodes().Get(context.Background(), name, options.GetOptions{})
+		if err != nil {
+			log.WithError(err).Warnf("Failed to get node, retrying")
+			time.Sleep(retrySleepTime)
+			continue
+		}
+		if calNode.Labels == nil {
+			calNode.Labels = map[string]string{}
+		}
+		if calNode.Annotations == nil {
+			calNode.Annotations = map[string]string{}
+		}
+
+		// Track if we need to perform an update.
+		needsUpdate := false
+
+		// Check if it has the annotation for k8s labels.
+		a, ok := calNode.Annotations[nodeLabelAnnotation]
+
+		// If there are labels present, then parse them. Otherwise this is
+		// a first-time sync, in which case there are no old labels.
+		var oldLabels map[string]string = map[string]string{}
+		if ok {
+			if err = json.Unmarshal([]byte(a), &oldLabels); err != nil {
+				log.WithError(err).Error("Failed to unmarshal node labels")
+				return
+			}
+		}
+		log.Debugf("Determined previously synced labels: %s", oldLabels)
+
+		// We've synced labels before. Determine diffs to apply.
+		// For each k/v in node.Labels, if it isn't present or the value
+		// differs, add it to the node.
+		for k, v := range node.Labels {
+			if v2, ok := calNode.Labels[k]; !ok || v != v2 {
+				log.Debugf("Adding node label %s=%s", k, v)
+				calNode.Labels[k] = v
+				needsUpdate = true
+			}
+		}
+
+		// For each k/v that used to be in the k8s node labels, but is no longer,
+		// remove it from the Calico node.
+		for k, v := range oldLabels {
+			if _, ok := node.Labels[k]; !ok {
+				// The old label is no longer present. Remove it.
+				log.Debugf("Deleting node label %s=%s", k, v)
+				delete(calNode.Labels, k)
+				needsUpdate = true
+			}
+		}
+
+		// Set the annotation to the correct values.
+		bytes, err := json.Marshal(node.Labels)
```

Comment:
i have a suspicion that it is impossible for `json.Marshal` to return an error here since `node.Labels` is a `map[string]string` and there no conceivable value we could set it to that would trigger an error.

would be okay to ignore the error in these kinds of cases?
reference: https://stackoverflow.com/questions/33903552/what-input-will-cause-golangs-json-marshal-to-return-an-error

Reply:
Yeah, it probably is. But, I think we have checks which will fail if we ignore errors. 

I don't mind having it there anyway, just in case....